File: ios-native/NeobankNative/Sources/Features/Kyc/KycController.swift
import Foundation
import Observation

@MainActor
@Observable
final class KycController {
    enum LoadState {
        case loading
        case loaded(KycStatusInfo)
        case failed(APIError)
    }

    private(set) var state: LoadState = .loading
    private(set) var hasSubmittedThisSession = false

    private let repository: KycRepository

    init(repository: KycRepository) {
        self.repository = repository
    }

    func load() async {
        state = .loading
        hasSubmittedThisSession = false
        do {
            state = .loaded(try await repository.status())
        } catch let error as APIError {
            state = .failed(error)
        } catch {
            state = .failed(.decoding())
        }
    }

    func submit(
        fullName: String,
        dateOfBirth: String,
        countryCode: String,
        documentType: String?,
        documentNumber: String?
    ) async throws {
        let result = try await repository.submit(
            fullName: fullName,
            dateOfBirth: dateOfBirth,
            countryCode: countryCode,
            documentType: documentType,
            documentNumber: documentNumber
        )
        hasSubmittedThisSession = true
        state = .loaded(KycStatusInfo(status: result.status, rejectionReason: result.rejectionReason))
    }
}

#if DEBUG
extension KycController {
    static var preview: KycController {
        KycController(repository: KycRepository(client: APIClient(baseURL: AppConfig.apiBaseURL)))
    }
}
#endif

File: android-native/app/src/main/kotlin/dev/horobets/neobanknative/features/kyc/KycController.kt
package dev.horobets.neobanknative.features.kyc

import androidx.compose.runtime.staticCompositionLocalOf
import dev.horobets.neobanknative.core.networking.ApiError
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow

sealed interface KycLoadState {
    data object Loading : KycLoadState
    data class Loaded(val info: KycStatusInfo) : KycLoadState
    data class Failed(val error: ApiError) : KycLoadState
}

class KycController(private val repository: KycRepository) {
    private val _state = MutableStateFlow<KycLoadState>(KycLoadState.Loading)
    val state: StateFlow<KycLoadState> = _state.asStateFlow()

    private val _hasSubmittedThisSession = MutableStateFlow(false)
    val hasSubmittedThisSession: StateFlow<Boolean> = _hasSubmittedThisSession.asStateFlow()

    suspend fun load() {
        _state.value = KycLoadState.Loading
        _hasSubmittedThisSession.value = false
        _state.value = try {
            KycLoadState.Loaded(repository.status())
        } catch (e: ApiError) {
            KycLoadState.Failed(e)
        } catch (e: Exception) {
            KycLoadState.Failed(ApiError.decoding())
        }
    }

    suspend fun submit(
        fullName: String,
        dateOfBirth: String,
        countryCode: String,
        documentType: String?,
        documentNumber: String?,
    ) {
        val result = repository.submit(fullName, dateOfBirth, countryCode, documentType, documentNumber)
        _hasSubmittedThisSession.value = true
        _state.value = KycLoadState.Loaded(KycStatusInfo(rawStatus = result.rawStatus, rejectionReason = result.rejectionReason))
    }
}

val LocalKycController = staticCompositionLocalOf<KycController> {
    error("No KycController provided — wrap the composition in AppEnvironment's providers")
}
